src/awkward/_v2/contents/listoffsetarray.py

```diff
@@ -1788,6 +1788,25 @@ def _to_arrow(self, pyarrow, mask_node, validbytes, length, options):
                 ),
             )
 
+    def _to_numpy(self, allow_missing):
+        if self.parameter("__array__") == "bytestring":
+            return self.nplike.array(
+                [
+                    ak._v2.behaviors.string.ByteBehavior(self[i]).__bytes__()
+                    for i in range(len(self))
+                ]
+            )
+
+        if self.parameter("__array__") == "string":
+            return self.nplike.array(
+                [
+                    ak._v2.behaviors.string.CharBehavior(self[i]).__str__()
+                    for i in range(len(self))
+                ]
+            )
```

Comment:
These are loops over all the elements in the array. Ideally, we'd pad them all to the maximum size (`pad_none` followed by `fill_none`) and take the padded buffer directly into NumPy in a vectorized operation, but I think `fill_none` isn't ready yet.

The next-best thing is this 40× faster `to_list` implementation; it iterates over all the elements in Python, making an intermediate Python list (as the original code did), but it does so only with built-in Python types, and that speeds it up a lot.

```suggestion
        array_param = self.parameter("__array__")
        if array_param == "bytestring" or array_param == "string":
            return self.nplike.array(self.to_list())
```